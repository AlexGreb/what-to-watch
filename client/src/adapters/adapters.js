export const createMovie = (movie) => {
  return {
    title: movie.name,
    genre: movie.genre,
    date: movie.released,
    backgroundColor: movie.backgroundColor,
    background: movie.backgroundImage,
    poster: movie.previewImage,
    posterImage: movie.posterImage,
    id: movie.id,
    description: movie.description,
    rating: movie.rating,
    votes: movie.scoresCount,
    director: movie.director,
    starring: movie.starring,
    runTime: movie.runTime,
    preview: movie.previewVideoLink,
    videoLink: movie.videoLink,
    isFavorite: movie.isFavorite,
  };
};

export const createReview = (review) => {
  return {
    id: review.id,
    name: review.user.name,
    userId: review.user.id,
    rating: review.rating,
    date: review.date,
    comment: review.comment
  };
};

const Rating = {
  AWESOME: 10,
  VERY_GOOD: 8,
  GOOD: 5,
  NORMAL: 3,
};

export const getRatingValue = (rating) => {
  if (rating === Rating.AWESOME) {
    return `Awesome`;
  }
  if (rating < Rating.AWESOME && rating >= Rating.VERY_GOOD) {
    return `Very good`;
  }
  if (rating < Rating.VERY_GOOD && rating >= Rating.GOOD) {
    return `Good`;
  }
  if (rating < Rating.GOOD && rating >= Rating.NORMAL) {
    return `Normal`;
  }
  return `Bad`;
};
