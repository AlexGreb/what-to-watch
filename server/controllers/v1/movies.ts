import type {ResponseData} from "../../types";
import type {IMovie} from "../../Models/movie";
import type {Request, Response} from "express";
import * as movieService from "../../services/v1/movieService";
import {getFavoriteMovies} from "../../services/v1/favoriteService";
import {getResError, setFavoriteMovies} from "../../utils"
import {validationResult} from "express-validator";

export const getAllMovies = async (req: Request, res: Response): Promise<void> => {
    try {
        const errors = validationResult(req);
        if(!errors.isEmpty()) {
            res.status(404).json({
                status: 'FAILED',
                data: {
                    error: errors,
                }
            })
        }
        const movies: IMovie[] = await movieService.getAllMovies();
        const favoriteMovies = await getFavoriteMovies(Number(req.params.userId));
        const moviesWithFavoriteUser = setFavoriteMovies(favoriteMovies, movies);
        const data: ResponseData<IMovie[]> = {
            status: 'OK',
            data: moviesWithFavoriteUser
        }
        res.json(data);
    } catch(error: unknown) {
        const {status, data} = getResError(error);
        res.status(status).json({
            status: 'FAILED',
            data
        });
    }
}

export const getMovie = async (req: Request, res: Response): Promise<void> => {
    const {params} = req;
    if(!params.movieId) return;
    try {
        const movie: IMovie | null = await movieService.getMovie(Number(params.movieId));

        if(movie) {
            res.status(200).json({
                status: 'OK',
                data: movie,
            })
        } else {
            res.status(404).json({
                status: 'FAILED',
                data: {
                    error: "Movie with this movieId wasn`t found",
                }
            })
        }
    } catch(error) {
        const {status, data} = getResError(error);
        res.status(status).json({
            status: 'FAILED',
            data
        });
    }
}

export const getPromo = async (req: Request, res: Response): Promise<void> =>{
    try {
        const promo: IMovie | null = await movieService.getPromo();
        res.json({
            status: 'OK',
            data: promo
        });
    } catch(error) {
        const {status, data} = getResError(error);
        res.status(status).json({
            status: 'FAILED',
            data
        });
    }
}
