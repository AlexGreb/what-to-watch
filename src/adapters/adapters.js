export const createMovie = (movie) => {
  return {
    title: movie.name,
    genre: movie.genre,
    date: movie.released,
    backgroundColor: movie.background_color,
    background: movie.background_image,
    poster: movie.preview_image,
    posterImage: movie.poster_image,
    id: movie.id,
    description: movie.description,
    rating: movie.rating,
    votes: movie.scores_count,
    director: movie.director,
    starring: movie.starring,
    runTime: movie.run_time,
    preview: movie.preview_video_link,
    videoLink: movie.video_link,
    isFavorite: movie.is_favorite,
  };
};

const Rating = {
  AWESOME: 10,
  VERY_GOOD: 8,
  GOOD: 5,
  NORMAL: 3,
};

export const getRatingValue = (rating) => {
  if (rating === Rating.AWESOME) {
    return `Awesome`;
  }
  if (rating < Rating.AWESOME && rating >= Rating.VERY_GOOD) {
    return `Very good`;
  }
  if (rating < Rating.VERY_GOOD && rating >= Rating.GOOD) {
    return `Good`;
  }
  if (rating < Rating.GOOD && rating >= Rating.NORMAL) {
    return `Normal`;
  }
  return `Bad`;
};
