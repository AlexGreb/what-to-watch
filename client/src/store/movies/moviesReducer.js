import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
import {createMovie, createReview} from '../../adapters/adapters';
import {dataUrl} from '../../constants/consts.js';
import {namespaces} from '../namespaces';

const initialState = {
  promoMovie: {},
  loadingMovies: false,
  loadingMyMovies: false,
  reviews: [],
  sendingReviewStatus: false,
  myListMovies: [],
  moviesList: [],
  isError: false
};

const Operations = {
  fetchMoviesList: createAsyncThunk(
      `${namespaces.MOVIES}/fetchMoviesList`,
      async (_, {extra}) => {
        const {data} = await extra.get(dataUrl.FILMS);
        return data.data.map(createMovie);
      }),
  fetchPromoMovie: createAsyncThunk(
      `${namespaces.MOVIES}/promo`,
      async (_, {extra}) => {
        const response = await extra.get(dataUrl.PROMO);
        return createMovie(response.data.data);
      }),
  fetchReviews: createAsyncThunk(
      `${namespaces.MOVIES}/fetchReviews`,
      async (movieId, {extra}) => {
        const response = await extra.get(`${dataUrl.REVIEWS}${movieId}`);
        return response.data.data.map(createReview);
      }
  ),
  fetchComment: createAsyncThunk(
      `${namespaces.MOVIES}/fetchComment`,
      async (reviewData, {extra}) => {
        const response = await extra.post(`${dataUrl.REVIEWS}${reviewData.movieId}`, reviewData.review);
        return response.data;
      }
  ),
  fetchMyListMovies: createAsyncThunk(
      `${namespaces.MOVIES}/fetchMyListMovies`,
      async (_, {extra}) => {
        const response = await extra.get(`${dataUrl.FAVORITE}`);
        return response.data.map(createMovie);
      }
  ),
  fetchChangeFavoriteStateMovie: createAsyncThunk(
      `${namespaces.MOVIES}/fetchChangeFavoriteStateMovie`,
      async (movie, {dispatch, fulfillWithValue, extra}) => {
        const response = await extra.post(`${dataUrl.FAVORITE}${movie.id}/${movie.isFavorite ? 0 : 1}`);
        dispatch(Operations.fetchMoviesList());
        dispatch(Operations.fetchPromoMovie());
        return fulfillWithValue(response.data);
      }
  )
};

const moviesSlice = createSlice({
  name: namespaces.MOVIES,
  initialState,
  reducers: {
    setSendingReviewStatus: (state, action) => {
      state.sendingReviewStatus = action.payload;
    },
  },
  extraReducers: {
    [Operations.fetchMoviesList.fulfilled]: (state, action) => {
      state.loadingMovies = false;
      state.moviesList = action.payload;
    },
    [Operations.fetchMoviesList.pending]: (state) => {
      state.loadingMovies = true;
    },
    [Operations.fetchMoviesList.rejected]: (state) => {
      state.isError = true;
    },
    [Operations.fetchPromoMovie.fulfilled]: (state, action) => {
      state.promoMovie = action.payload;
    },
    [Operations.fetchPromoMovie.rejected]: (state) => {
      state.isError = true;
    },
    [Operations.fetchReviews.fulfilled]: (state, action) => {
      state.reviews = action.payload;
    },
    [Operations.fetchComment.fulfilled]: (state) => {
      state.sendingReviewStatus = true;
    },
    [Operations.fetchComment.rejected]: (state) => {
      state.sendingReviewStatus = false;
    },
    [Operations.fetchMyListMovies.fulfilled]: (state, action) => {
      state.loadingMyMovies = false;
      state.myListMovies = action.payload;
    },
    [Operations.fetchMyListMovies.pending]: (state) => {
      state.loadingMyMovies = true;
    },
    [Operations.fetchMyListMovies.rejected]: (state) => {
      state.loadingMyMovies = false;
      state.isError = true;
    }
  }
});

export const {setSendingReviewStatus} = moviesSlice.actions;
export {
  Operations,
  initialState
};
export default moviesSlice.reducer;

