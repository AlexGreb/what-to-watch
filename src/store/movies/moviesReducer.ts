import {createSlice, createAsyncThunk, PayloadAction} from '@reduxjs/toolkit';
import {createMovie, createReview} from '../../adapters/adapters';
import {AppRoute} from '../../constants/consts.ts';
import {Namespaces} from '../storeNamespaces';
import {Movie, MovieReviews} from '../../types/movie';

type InitialState {
  promoMovie?: Movie | {},
  loadingMovies: boolean,
  loadingMyMovies: boolean,
  reviews: MovieReviews | [],
  sendingReviewStatus: boolean,
  myListMovies: Array<Movie> | [],
  moviesList: Array<Movie> | [],
  isError: boolean
}

const initialState:InitialState = {
  promoMovie: {},
  loadingMovies: false,
  loadingMyMovies: false,
  reviews: [],
  sendingReviewStatus: false,
  myListMovies: [],
  moviesList: [],
  isError: false
};

const Operations = {
  fetchMoviesList: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchMoviesList`,
      async (_, {extra}) => {
        const response = await extra.get(AppRoute.FILMS);
        return response.data.map(createMovie);
      }),
  fetchPromoMovie: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchPromoMovie`,
      async (_, {extra}) => {
        const response = await extra.get(AppRoute.PROMO);
        const a = await response.json()
        return createMovie(response.data);
      }),
  fetchReviews: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchReviews`,
      async (movieId: number, {extra}) => {
        const response = await extra.get(`${AppRoute.REVIEWS}${movieId}`);
        return response.data.map(createReview);
      }
  ),
  fetchComment: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchComment`,
      async (reviewData, {extra}) => {
        const response = await extra.post(`${AppRoute.REVIEWS}${reviewData.movieId}`, reviewData.review);
        return response.data;
      }
  ),
  fetchMyListMovies: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchMyListMovies`,
      async (_, {extra}) => {
        const response = await extra.get(`${AppRoute.FAVORITE}`);
        return response.data.map(createMovie);
      }
  ),
  fetchChangeFavoriteStateMovie: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchChangeFavoriteStateMovie`,
      async (movie: Movie, {dispatch, fulfillWithValue, extra}) => {
        const response = await extra.post(`${AppRoute.FAVORITE}${movie.id}/${movie.isFavorite ? 0 : 1}`);
        dispatch(Operations.fetchMoviesList());
        dispatch(Operations.fetchPromoMovie());
        return fulfillWithValue(response.data);
      }
  )
};

const moviesSlice = createSlice({
  name: Namespaces.MOVIES,
  initialState,
  reducers: {
    setSendingReviewStatus: (state, action) => {
      state.sendingReviewStatus = action.payload;
    },
  },
  extraReducers: {
    [Operations.fetchMoviesList.fulfilled.type]: (state, action) => {
      state.loadingMovies = false;
      state.moviesList = action.payload;
    },
    [Operations.fetchMoviesList.pending.type]: (state) => {
      state.loadingMovies = true;
    },
    [Operations.fetchMoviesList.rejected.type]: (state) => {
      state.isError = true;
    },
    [Operations.fetchPromoMovie.fulfilled.type]: (state, action) => {
      state.promoMovie = action.payload;
    },
    [Operations.fetchPromoMovie.rejected.type]: (state) => {
      state.isError = true;
    },
    [Operations.fetchReviews.fulfilled.type]: (state, action) => {
      state.reviews = action.payload;
    },
    [Operations.fetchComment.fulfilled.type]: (state) => {
      state.sendingReviewStatus = true;
    },
    [Operations.fetchComment.rejected.type]: (state) => {
      state.sendingReviewStatus = false;
    },
    [Operations.fetchMyListMovies.fulfilled.type]: (state, action) => {
      state.loadingMyMovies = false;
      state.myListMovies = action.payload;
    },
    [Operations.fetchMyListMovies.pending.type]: (state) => {
      state.loadingMyMovies = true;
    },
    [Operations.fetchMyListMovies.rejected.type]: (state) => {
      state.loadingMyMovies = false;
      state.isError = true;
    }
  }
});

export const {setSendingReviewStatus} = moviesSlice.actions;
export {
  Operations,
  initialState
};
export default moviesSlice.reducer;

