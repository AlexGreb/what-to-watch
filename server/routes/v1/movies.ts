import {Router} from "express";
import {getAllMovies, getMovie, getPromo} from "../../controllers/v1/movies";
import {getMovieReviews} from "../../controllers/v1/reviews";
import {authMiddleware} from "../../middlewares/authMiddleware";
import {query} from "express-validator";
import config from "config";

const moviesRouter = Router();
const apiUrl = config.get('apiUrl');
console.log(apiUrl);

moviesRouter.get(`${apiUrl}movies`, [
    authMiddleware,
    query('userId', 'userId is required').notEmpty()
], getAllMovies);

moviesRouter.get(`${apiUrl}movies/:movieId`,[authMiddleware], getMovie);

moviesRouter.get(`${apiUrl}promo`,[authMiddleware], getPromo);

moviesRouter.get(`${apiUrl}reviews/:movieId`, [authMiddleware], getMovieReviews);


export {moviesRouter};
