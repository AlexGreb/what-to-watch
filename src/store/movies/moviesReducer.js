import {createSlice} from '@reduxjs/toolkit';
import {moviesList} from '../../mocks/movies.js';
import {createMovie} from '../../adapters/adapters';

const initialState = {
  moviesList: []
};

const moviesSlice = createSlice({
  name: `movies`,
  initialState,
  reducers: {
    setMoviesList: (state, action) => {
      state.moviesList = action.payload;
    }

  }
});

const {setMoviesList} = moviesSlice.actions;

const Operations = {
  getMoviesList: () => (dispatch) => {
    setTimeout(() => {
      dispatch(setMoviesList(moviesList.map(createMovie)));
    }, 1000);
  }
};

export {setMoviesList, Operations, initialState};
export default moviesSlice.reducer;

