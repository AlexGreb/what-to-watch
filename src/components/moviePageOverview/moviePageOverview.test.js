/**
 * @jest-environment jsdom
 */
import React from 'react';
import {render} from '@testing-library/react';
import MoviePageOverview from './moviePageOverview.jsx';
import {movie} from '../../mocks/movies.js';
import {createMovie} from '../../adapters/adapters.js';

describe(`MoviePageDetails`, () => {
  it(`Should render correctly a MoviePageDetails component`, () => {
    const {asFragment} = render(<MoviePageOverview movie={createMovie(movie)}/>);
    expect(asFragment()).toMatchSnapshot();
  });

});

