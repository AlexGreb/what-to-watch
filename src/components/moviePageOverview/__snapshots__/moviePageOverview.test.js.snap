// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePageDetails Should render correctly a MoviePageDetails component 1`] = `
<DocumentFragment>
  <div
    class="movie-rating"
  >
    <div
      class="movie-rating__score"
    >
      4.5
    </div>
    <p
      class="movie-rating__meta"
    >
      <span
        class="movie-rating__level"
      >
        Normal
      </span>
      <span
        class="movie-rating__count"
      >
        1635992 ratings
      </span>
    </p>
  </div>
  <div
    class="movie-card__text"
  >
    The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.
    <p
      class="movie-card__director"
    >
      <strong>
        Director: Quentin Tarantino
      </strong>
    </p>
    <p
      class="movie-card__starring"
    >
      <strong>
        Starring: John Travolta, Uma Thurman, Samuel L. Jackson
      </strong>
    </p>
  </div>
</DocumentFragment>
`;
